freestyle/shared/src/main/scala/freestyle/module.scala

```diff
@@ -16,171 +16,190 @@
+      cs.baseClasses.exists(_.name == TypeName("FreeModuleLike") )
 
-    def filterEffectVals(trees: Template): List[ValDef]  =
-      trees.collect { case v: ValDef if v.mods.hasFlag(Flag.DEFERRED) => v }
+    def expandClass(cs: ClassSymbol): List[Type] =
+      cs.info.decls.toList
+        .collect { case met: MethodSymbol if met.isAbstract => met }
+        .flatMap(x => expandEffectType(x.returnType))
 
-    val LL = freshTypeName("LL$")
-
-    def toImplArg(effVal: ValDef): ValDef = effVal match {
-      case q"$mods val $name: $eff[..$args]" =>
-        q"$mods val $name: $eff[$LL, ..$args]"
+    tpe.typeSymbol match {
+      case cs: ClassSymbol =>
+        if (isModuleFS(cs)) expandClass(cs) else List(tpe.typeConstructor)
+      case _ => Nil
     }
+  }
 
-    def mkModuleTrait(cls: ClassDef): ClassDef = {
-      val FF = freshTypeName("FF$")
-      // this is to make a TypeDef for `$FF[_]`
-      val wildcard = TypeDef(Modifiers(Flag.PARAM), typeNames.WILDCARD, List(), TypeBoundsTree(EmptyTree, EmptyTree))
-      val ffTParam = TypeDef(Modifiers(Flag.PARAM), FF, List(wildcard), TypeBoundsTree(EmptyTree, EmptyTree))
-
-      val ClassDef(mods, name, tparams, Template(parents, self, body)) = cls
-
-      val nbody = body.map {
-        case q"$mods val $name: $eff[..$args]" => q"$mods val $name: $eff[$FF, ..$args]"
-        case x => x
+  private[this] def typeToTypeTree(tpe: Type): ValidatedNel[String, Tree] =
+    Validated.catchNonFatal(c.parse(tpe.toString)).leftMap(t =>
+      NonEmptyList.of(s"unable to convert type $tpe to tree"))
+
+  private[this] lazy val defaultf0: ValidatedNel[String, Tree] =
+    q"""type Op[$AA] = Nothing""".validNel
+  private[this] def defaultf1(tpt: Tree): ValidatedNel[String, Tree] =
+    q"""type Op[$AA] = $tpt.Op[$AA]""".validNel
+
+  private[this] def makeCatsCoproduct(
+    tpts: List[Tree]
+  ): ValidatedNel[String, Tree] = tpts match {
+    case Nil                    => defaultf0
+    case tpt :: Nil             => defaultf1(tpt)
+    case head1 :: head2 :: tail =>
+
+      val max = tail.length
+      def Op(i: Int) = (max - i - 1) match {
+        case 0 => TypeName("Op")
+        case n => TypeName(s"Op$n")
       }
-      val nimpl = Template(parents :+ tq"freestyle.FreeModuleLike", self, nbody)
-      ClassDef(mods, name, ffTParam :: tparams, nimpl)
-    }
-
-    def mkModuleObject(userTrait: ClassDef): ModuleDef = {
-      val mod = userTrait.name
-      val tts = userTrait.tparams
-      val tns = tts.map(_.name)
-      val AA = freshTypeName("AA$")
-      val ev = freshTermName("ev$")
-      val xx = freshTermName("xx$")
-
-      val effArgs: List[ValDef] = filterEffectVals(userTrait.impl).map( v => toImplArg(v) )
 
-      q"""
-        object ${mod.toTermName} {
+      tail.zipWithIndex.foldLeft(
+        q"""type ${Op(-1)}[$AA] = $Coproduct[$head2.Op, $head1.Op, $AA]""": Tree
+      ) { (prefix, tup) =>
+        val (tpt, i) = tup
+        q"""
+          ..$prefix
```

Comment:
Since you are not getting any data out from the `prefix`, it seems to me you can just use `map` here.
If you want to return a single `Tree`, you can do the unquoting at the end. However, it may be more informative to return a `List[T]`, where `T` is the actual _type_ in `scala.reflect` of what a member type definition is.